src/server/pps/server/api_server.go

```diff
@@ -2944,42 +2938,41 @@ func (a *apiServer) DeleteAll(ctx context.Context, request *types.Empty) (respon
 }
 
 // ActivateAuth implements the protobuf pps.ActivateAuth RPC
-func (a *apiServer) ActivateAuth(ctx context.Context, req *pps.ActivateAuthRequest) (resp *pps.ActivateAuthResponse, retErr error) {
+func (a *apiServer) ActivateAuth(ctx context.Context, req *pps.ActivateAuthRequest) (*pps.ActivateAuthResponse, error) {
+	var resp *pps.ActivateAuthResponse
+	if err := a.txnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+		var err error
+		resp, err = a.ActivateAuthInTransaction(ctx, txnCtx, req)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (a *apiServer) ActivateAuthInTransaction(ctx context.Context, txnCtx *txncontext.TransactionContext, req *pps.ActivateAuthRequest) (resp *pps.ActivateAuthResponse, retErr error) {
 	// Unauthenticated users can't create new pipelines or repos, and users can't
 	// log in while auth is in an intermediate state, so 'pipelines' is exhaustive
-	var eg errgroup.Group
-	pipelineInfo := &pps.PipelineInfo{}
-	if err := a.pipelines.ReadOnly(ctx).List(pipelineInfo, col.DefaultOptions(), func(string) error {
-		pipeline := proto.Clone(pipelineInfo).(*pps.PipelineInfo)
-		pipelineName := pipeline.Pipeline.Name
+	pi := &pps.PipelineInfo{}
+	if err := a.pipelines.ReadOnly(ctx).List(pi, col.DefaultOptions(), func(string) error {
 		// 1) Create a new auth token for 'pipeline' and attach it, so that the
 		// pipeline can authenticate as itself when it needs to read input data
-		eg.Go(func() error {
-			return a.txnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
-				token, err := a.env.AuthServer.GetPipelineAuthTokenInTransaction(txnCtx, pipelineName)
-				if err != nil {
-					return errors.Wrapf(grpcutil.ScrubGRPC(err), "could not generate pipeline auth token")
-				}
-
-				pipelineInfo := &pps.PipelineInfo{}
-				if err := a.updatePipeline(txnCtx, pipelineName, pipelineInfo, func() error {
-					pipelineInfo.AuthToken = token
-					return nil
-				}); err != nil {
-					return errors.Wrapf(err, "could not update \"%s\" with new auth token", pipelineName)
-				}
-				return nil
-			})
-		})
+		token, err := a.env.AuthServer.GetPipelineAuthTokenInTransaction(txnCtx, pi.Pipeline.Name)
+		if err != nil {
+			return errors.Wrapf(grpcutil.ScrubGRPC(err), "could not generate pipeline auth token")
+		}
+		if err := a.updatePipeline(txnCtx, pi.Pipeline.Name, pi, func() error {
+			pi.AuthToken = token
+			return nil
+		}); err != nil {
+			return errors.Wrapf(err, "could not update \"%s\" with new auth token", pi.Pipeline.Name)
+		}
 		// put 'pipeline' on relevant ACLs
-		if err := a.fixPipelineInputRepoACLs(ctx, pipeline, nil); err != nil {
-			return err
+		if err := a.fixPipelineInputRepoACLsInTransaction(txnCtx, pi, nil); err != nil {
+			return errors.Wrapf(err, "fix repo ACLs for pipeline %q", pi.Pipeline.Name)
```

Comment:
It would be nice to log that we've done all of this, but to some extent the logs would have to be done after the transaction commits to be strictly correct.  Maybe DEBUG-level log each pipeline that was updated, and the overall outcome of the transaction a couple levels up?

Reply:
I kind of like the idea of a per-transaction log buffer, but just DEBUG logging everything seems clearer. We lost a lot of internal info when we moved over to everything internal being a transaction, it would be nice to recover some of it.